File: src/main/java/com/yupi/template/service/ImageServiceStrategy.java
package com.yupi.template.service;

import com.yupi.template.model.dto.image.ImageRequest;
import com.yupi.template.model.enums.ImageMethodEnum;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import jakarta.annotation.PostConstruct;
import jakarta.annotation.Resource;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;

/**
 * 图片服务策略选择器
 * 根据图片来源类型选择对应的图片服务实现
 * 
 * 设计说明：
 * - 自动注册所有 ImageSearchService 实现
 * - 根据 ImageMethodEnum 的元数据自动选择正确的参数
 * - 支持服务可用性检查和自动降级
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
@Service
@Slf4j
public class ImageServiceStrategy {

    @Resource
    private List<ImageSearchService> imageSearchServices;

    /**
     * 图片服务映射：ImageMethodEnum -> ImageSearchService
     */
    private final Map<ImageMethodEnum, ImageSearchService> serviceMap = new EnumMap<>(ImageMethodEnum.class);

    @PostConstruct
    public void init() {
        // 将所有 ImageSearchService 实现注册到映射表
        for (ImageSearchService service : imageSearchServices) {
            ImageMethodEnum method = service.getMethod();
            serviceMap.put(method, service);
            log.info("注册图片服务: {} -> {} (AI生图: {}, 降级: {})", 
                    method.getValue(), 
                    service.getClass().getSimpleName(),
                    method.isAiGenerated(),
                    method.isFallback());
        }
    }

    /**
     * 根据图片请求获取图片（推荐方法）
     *
     * @param imageSource 图片来源
     * @param request     图片请求对象
     * @return 图片获取结果
     */
    public ImageResult getImage(String imageSource, ImageRequest request) {
        ImageMethodEnum method = resolveMethod(imageSource);
        ImageSearchService service = serviceMap.get(method);
        
        if (service == null || !service.isAvailable()) {
            log.warn("图片服务不可用: {}, 尝试降级", method);
            return handleFallback(request.getPosition());
        }

        String imageUrl = service.getImage(request);
        
        if (imageUrl != null && !imageUrl.isEmpty()) {
            return new ImageResult(imageUrl, method);
        } else {
            log.warn("图片获取失败, 使用降级方案, method={}", method);
            return handleFallback(request.getPosition());
        }
    }

    /**
     * 根据图片来源获取对应的图片（兼容旧接口）
     *
     * @param imageSource 图片来源（PEXELS / NANO_BANANA 等）
     * @param keywords    关键词（用于图库检索）
     * @param prompt      提示词（用于 AI 生图）
     * @return 图片获取结果
     */
    public ImageResult getImage(String imageSource, String keywords, String prompt) {
        ImageRequest request = ImageRequest.builder()
                .keywords(keywords)
                .prompt(prompt)
                .build();
        return getImage(imageSource, request);
    }

    /**
     * 解析图片来源，处理未知值
     */
    private ImageMethodEnum resolveMethod(String imageSource) {
        ImageMethodEnum method = ImageMethodEnum.getByValue(imageSource);
        if (method == null) {
            log.warn("未知的图片来源: {}, 默认使用 {}", imageSource, ImageMethodEnum.getDefaultSearchMethod());
            return ImageMethodEnum.getDefaultSearchMethod();
        }
        return method;
    }

    /**
     * 处理降级逻辑
     */
    private ImageResult handleFallback(Integer position) {
        int pos = position != null ? position : 1;
        String fallbackUrl = getFallbackImage(pos);
        return new ImageResult(fallbackUrl, ImageMethodEnum.getFallbackMethod());
    }

    /**
     * 获取指定方法的图片服务
     *
     * @param method 图片方法
     * @return 图片服务，未找到返回 null
     */
    public ImageSearchService getService(ImageMethodEnum method) {
        return serviceMap.get(method);
    }

    /**
     * 获取降级图片
     *
     * @param position 位置序号
     * @return 降级图片 URL
     */
    public String getFallbackImage(int position) {
        // 优先使用已注册服务的降级方案
        ImageSearchService defaultService = serviceMap.get(ImageMethodEnum.getDefaultSearchMethod());
        if (defaultService != null) {
            return defaultService.getFallbackImage(position);
        }
        return String.format("https://picsum.photos/800/600?random=%d", position);
    }

    /**
     * 获取所有已注册的图片服务类型
     */
    public List<ImageMethodEnum> getRegisteredMethods() {
        return List.copyOf(serviceMap.keySet());
    }

    /**
     * 图片获取结果
     */
    public static class ImageResult {
        private final String url;
        private final ImageMethodEnum method;

        public ImageResult(String url, ImageMethodEnum method) {
            this.url = url;
            this.method = method;
        }

        public String getUrl() {
            return url;
        }

        public ImageMethodEnum getMethod() {
            return method;
        }

        public boolean isSuccess() {
            return url != null && !url.isEmpty();
        }
    }
}

File: python-backend/app/services/image_service_strategy.py
"""图片服务策略选择器（第 5 期新增）"""

import logging
from typing import Dict, List, Optional

from app.models.enums import ImageMethodEnum
from app.services.image_search_service import ImageSearchService
from app.services.pexels_service import PexelsService
from app.services.nano_banana_service import NanoBananaService
from app.services.mermaid_service import MermaidService
from app.services.iconify_service import IconifyService
from app.services.emoji_pack_service import EmojiPackService
from app.services.svg_diagram_service import SvgDiagramService
from app.services.cos_service import CosService
from app.schemas.image import ImageRequest, ImageData
from app.constants.article import ArticleConstant

logger = logging.getLogger(__name__)


class ImageResult:
    """图片获取结果"""
    
    def __init__(self, url: str, method: ImageMethodEnum):
        self.url = url
        self.method = method
    
    def is_success(self) -> bool:
        """判断是否成功"""
        return self.url is not None and len(self.url) > 0


class ImageServiceStrategy:
    """
    图片服务策略选择器
    根据图片来源类型选择对应的图片服务实现
    """
    
    def __init__(self):
        # 初始化所有图片服务
        self.service_map: Dict[ImageMethodEnum, ImageSearchService] = {}
        self.cos_service = CosService()
        
        # 注册所有服务
        self._register_services()
    
    def _register_services(self):
        """注册所有图片服务"""
        services = [
            PexelsService(),
            NanoBananaService(),
            MermaidService(),
            IconifyService(),
            EmojiPackService(),
            SvgDiagramService(),
        ]
        
        for service in services:
            method = service.get_method()
            self.service_map[method] = service
            logger.info(
                f"注册图片服务: {method.value} -> {service.__class__.__name__} "
                f"(AI生图: {method.is_ai_generated()}, 降级: {method.is_fallback()})"
            )
    
    async def get_image_and_upload(
        self,
        image_source: str,
        request: ImageRequest
    ) -> ImageResult:
        """
        获取图片并上传到 COS（推荐方法）
        统一处理所有图片来源的上传逻辑
        
        Args:
            image_source: 图片来源
            request: 图片请求对象
            
        Returns:
            图片获取结果（包含 COS URL）
        """
        method = self._resolve_method(image_source)
        service = self.service_map.get(method)
        
        if service is None or not service.is_available():
            logger.warning(f"图片服务不可用: {method}, 尝试降级")
            return await self._handle_fallback_with_upload(request.position)
        
        try:
            # 1. 获取图片数据
            image_data = await service.get_image_data(request)
            
            if image_data is None or not image_data.is_valid():
                logger.warning(f"图片数据获取失败, 使用降级方案, method={method}")
                return await self._handle_fallback_with_upload(request.position)
            
            # 2. 上传到 COS
            folder = self._get_folder_for_method(method)
            cos_url = await self.cos_service.upload_image_data(image_data, folder)
            
            if cos_url:
                logger.info(f"图片获取并上传成功, method={method}, cosUrl={cos_url}")
                return ImageResult(cos_url, method)
            else:
                logger.warning(f"图片上传 COS 失败, 使用降级方案, method={method}")
                return await self._handle_fallback_with_upload(request.position)
        except Exception as e:
            logger.error(f"获取图片并上传异常, method={method}, error={e}")
            return await self._handle_fallback_with_upload(request.position)
    
    def _resolve_method(self, image_source: str) -> ImageMethodEnum:
        """解析图片来源，处理未知值"""
        try:
            return ImageMethodEnum(image_source)
        except ValueError:
            logger.warning(
                f"未知的图片来源: {image_source}, "
                f"默认使用 {ImageMethodEnum.get_default_search_method().value}"
            )
            return ImageMethodEnum.get_default_search_method()
    
    async def _handle_fallback_with_upload(self, position: Optional[int]) -> ImageResult:
        """处理降级逻辑"""
        pos = position if position else 1
        fallback_url = self._get_fallback_image(pos)
        
        # 将降级图片也上传到 COS
        fallback_data = ImageData.from_url(fallback_url)
        cos_url = await self.cos_service.upload_image_data(fallback_data, "fallback")
        
        # 如果上传失败，直接使用原始 URL
        final_url = cos_url if cos_url else fallback_url
        return ImageResult(final_url, ImageMethodEnum.get_fallback_method())
    
    def _get_folder_for_method(self, method: ImageMethodEnum) -> str:
        """根据图片方法获取 COS 文件夹"""
        folder_map = {
            ImageMethodEnum.PEXELS: "pexels",
            ImageMethodEnum.NANO_BANANA: "nano-banana",
            ImageMethodEnum.MERMAID: "mermaid",
            ImageMethodEnum.ICONIFY: "iconify",
            ImageMethodEnum.EMOJI_PACK: "emoji-pack",
            ImageMethodEnum.SVG_DIAGRAM: "svg-diagram",
            ImageMethodEnum.PICSUM: "picsum",
        }
        return folder_map.get(method, "unknown")
    
    def _get_fallback_image(self, position: int) -> str:
        """获取降级图片"""
        # 优先使用已注册服务的降级方案
        default_service = self.service_map.get(ImageMethodEnum.get_default_search_method())
        if default_service:
            return default_service.get_fallback_image(position)
        return ArticleConstant.PICSUM_URL_TEMPLATE.format(position)
    
    def get_registered_methods(self) -> List[ImageMethodEnum]:
        """获取所有已注册的图片服务类型"""
        return list(self.service_map.keys())
